dali/kernels/imgproc/sampler.h

```diff
@@ -0,0 +1,183 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_IMGPROC_SAMPLER_H_
+#define DALI_KERNELS_IMGPROC_SAMPLER_H_
+
+#include <cmath>
+#include "dali/common.h"
+#include "dali/kernels/common/convert.h"
+#include "dali/kernels/tensor_view.h"
+#include "dali/kernels/imgproc/surface.h"
+
+namespace dali {
+namespace kernels {
+
+template <DALIInterpType interp, typename In>
+struct Sampler;
+
+template <DALIInterpType interp, typename T>
+__host__ __device__ Sampler<interp, T> make_sampler(const Surface2D<T> &surface) {
+  return Sampler<interp, T>(surface);
+}
+
+template <typename In>
+struct Sampler<DALI_INTERP_NN, In> {
+  Sampler() = default;
+  __host__ __device__ explicit Sampler(const Surface2D<const In> &surface)
+  : surface(surface) {}
+
+  Surface2D<const In> surface;
+
+  template <typename T = In>
+  __host__ __device__ T at(
+      int x, int y, int c,
+      const In *border_value) const {
+    if (x < 0 || x >= surface.width ||
+        y < 0 || y >= surface.height) {
+      return clamp<T>(border_value[c]);
+    } else {
+      return clamp<T>(surface(x, y, c));
+    }
+  }
+
+  template <typename T = In>
+  __host__ __device__ T at(
+      float x, float y, int c,
+      const In *border_value) const {
+  }
+
+  template <typename T>
+  __host__ __device__ void
+  operator()(T *pixel, int x, int y, int c, const In *border_value) const {
```

Comment:
First parameter of these operators sometimes is called `pixel` and sometimes `out_pixel`. Maybe it could be uniform across all of them. Unless there is some significance to it, which I'm missing.

Reply:
No, it's always an output.